Mage/src/main/java/mage/abilities/effects/common/MeldEffect.java

```diff
@@ -46,48 +50,58 @@ public MeldEffect copy() {
-                    meldCard.setOwnerId(controller.getId());
-                    meldCard.setTopHalfCard(meldWithCard, game);
-                    meldCard.setBottomHalfCard(sourceCard, game);
-                    meldCard.setMelded(true, game);
-                    game.addMeldCard(meldCard.getId(), meldCard);
-                    game.getState().addCard(meldCard);
-                    meldCard.setZone(Zone.EXILED, game);
-                    controller.moveCards(meldCard, Zone.BATTLEFIELD, source, game);
-                }
-                return true;
-            }
+        Permanent sourcePermanent = source.getSourcePermanentIfItStillExists(game);
+        if (controller == null
+                || sourcePermanent == null
+                || !sourcePermanent.isControlledBy(controller.getId())
+                || !sourcePermanent.isOwnedBy(controller.getId())) {
+            return false;
+        }
+        // Find the two permanents to meld.
+        FilterPermanent filter = new FilterControlledPermanent("permanent named " + meldWithName);
+        filter.add(new NamePredicate(meldWithName));
+        filter.add(TargetController.YOU.getOwnerPredicate());
+        if (!game.getBattlefield().contains(filter, source, game, 1)) {
+            return false;
+        }
+        TargetPermanent target = new TargetPermanent(filter);
+        target.setNotTarget(true);
+        controller.choose(outcome, target, source, game);
+
+        Permanent meldWithPermanent = game.getPermanent(target.getFirstTarget());
+        if (sourcePermanent == null || meldWithPermanent == null) {
+            return false;
+        }
+        Cards cards = new CardsImpl(sourcePermanent);
+        cards.add(meldWithPermanent);
+        controller.moveCards(cards, Zone.EXILED, source, game);
+        // Create the meld card and move it to the battlefield.
+        Card sourceCard = cards.get(sourcePermanent.getId(), game);
+        Card meldWithCard = cards.get(meldWithPermanent.getId(), game);
+        if (sourceCard == null
+                || meldWithCard == null
+                || !sourceCard.meldsWith(meldWithCard)
+                || !meldWithCard.meldsWith(sourceCard)) {
+            return true;
+        }
+        List<CardInfo> cardInfoList = CardRepository.instance.findCards(
+                new CardCriteria()
+                        .name(meldIntoName)
+                        .setCodes(sourceCard.getExpansionSetCode())
+                        .nightCard(true)
+        );
+        if (cardInfoList.isEmpty()) {
+            return false;
         }
-        return false;
+        MeldCard meldCard = (MeldCard) cardInfoList.get(0).getCard().copy();
+        meldCard.setOwnerId(controller.getId());
+        meldCard.setTopHalfCard(meldWithCard, game);
+        meldCard.setBottomHalfCard(sourceCard, game);
+        meldCard.setMelded(true, game);
+        game.addMeldCard(meldCard.getId(), meldCard);
+        game.getState().addCard(meldCard);
```

Comment:
Why isn't this part of `game.addMeldCard()`?

Reply:
This and the above comments fall into the category of things I'm not trying to change about the meld implementation because that's a whole separate task to me.